app/models/room.rb

```diff
@@ -16,7 +16,13 @@ class Room < ApplicationRecord
   validates :meeting_id, presence: true, uniqueness: true
 
   before_validation :set_friendly_id, :set_meeting_id, on: :create
+
   after_create :create_meeting_options
+  after_create :generate_viewer_access_code, if: :auto_generate_viewer_access_code?
+  after_create :generate_moderator_access_code, if: :auto_generate_moderator_access_code?
+
+  after_find :generate_viewer_access_code, if: :auto_generate_viewer_access_code?
+  after_find :generate_moderator_access_code, if: :auto_generate_moderator_access_code?
```

Comment:
Could we use `after_initialize` instead of both callbacks?

> The [after_initialize](https://api.rubyonrails.org/v7.0.3.1/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_initialize) callback will be called whenever an Active Record object is instantiated, either by directly using new or when a record is loaded from the database.

Reply:
Here's an inclusive response hope that you don't get bored 🤣 

The problem is due to the nature of `after_initialize`.
The callback will be called whenever a new instance of the Room model is created, wither after a fetching operation or when attempting to create a new record.
When attempting to create a new record the `after_initialize` will be one of the first if not the first callback to be called for a record that still not `:persisted?` in the DB.

The access codes are room meeting options records where each refers to a room and a meeting option.
Auto generating the access codes is part of the fallowing pipeline:
1. Room model gets instantiated and populated.
2. Room gets its unique meeting and friendly ids set -> `before_validation :set_friendly_id, :set_meeting_id, on: :create`.
3. Room record gets persisted to the DB.
4. Room meeting options records gets auto generated and persisted to the DB with their corresponding default values -> `after_create :create_meeting_options`.
5. Room access codes meeting options will **update** their default values with an auto generated codes for certain scenarios.

So, with such pipeline the access codes auto generation on create requires having the room meeting options records in place, which in turn requires having the room record itself persisted in the DB.

That was the reasoning behind avoiding the `after_initialize` since it will cause us additional queries without any benefit (**performance loss**).

What we can do instead is to leave only the `after_find` callback which will give the expected outcome but that means in some expected **frequent** cases additional waiting time on any feature in GL3 that fetches a room record (like the join page, rooms page, room page, admin server rooms,... ) and that's why we wanted to push that waiting time whenever possible to the creation process for a better UX.